### PR TITLE
add code upload and encrypt payload here instead of in ironcli

### PR DESCRIPTION
**Pull request by rdallman**
the upload and encrypt things could be handy for certain programs.. last thing left is docker login stuff but I'm not touching that until we get a certain swapi pr in.

### CODE REVIEW

**Comment by rdallman** [comment]
this is mostly copied over from the deleted code [here](https://github.com/iron-io/ironcli/issues/78) so should be pretty straight forward

File: api/api.go [hunk 1]
```diff
@@ -138,39 +152,49 @@ func (u *URL) Req(method string, in, out interface{}) error {
...
 	}
 
-	request.ContentLength = int64(len(body))
+	// bytes.Buffer and bytes.Reader both implement `Len() int`, zip uses former, all else latter.
+	// if this changes for some reason, looky here
+	if s, ok := body.(interface {
+		Len() int
+	}); ok {
+		request.ContentLength = int64(s.Len())
+	}
```
Comment (nikhilm) [inline api/api.go]: Nit: Do we want to panic/error in the !ok case?

File: worker/methods.go [hunk 2]
```diff
@@ -148,6 +161,73 @@ func (w *Worker) CodePackageRevisions(codeId string) (code Code, err error) {
...
+	mMetaWriter, err := mWriter.CreateFormField("data")
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.NewEncoder(mMetaWriter).Encode(args); err != nil {
+		return nil, err
+	}
+
+	if zipName != "" {
```
Comment (nikhilm) [inline worker/methods.go]: Why do we open one zip and write to another zip instead of just sending the bytes of the zip file?

Comment (nikhilm) [inline worker/methods.go]: I know you just copied the code over, but something to fix later?

Comment (rdallman) [inline worker/methods.go]: it's multipart...

File: worker/methods.go [hunk 3]
```diff
@@ -361,3 +441,66 @@ func (w *Worker) ScheduleCancel(scheduleId string) (err error) {
...
+//   of the modulus of the given RSA key first bits (i.e. 2048 = first 256 bytes), followed by
+//   the aes cipher with a new, random iv in the first 12 bytes,
+//   and the auth tag in the last 16 bytes of the cipher.
+// - must have RSA key >= 1024
+// - end format w/ RSA of 2048 for display purposes, all base64 encoded:
+//   [ 256 byte RSA encrypted AES key | len(payload) AES-GCM cipher | 16 bytes AES tag | 12 bytes AES nonce ]
+// - each task will be encrypted with a different AES session key
+//
+// EncryptPayloads will return a copy of the input tasks with the Payload field modified
+// to be encrypted as described above. Upon any error, the tasks returned will be nil.
```
Comment (nikhilm) [inline worker/methods.go]: Add a link to that README describing the encryption?

File: api/api.go (hunk 1 above)
Comment (rdallman) [inline api/api.go]: no, content length isn't a strict requirement

File: worker/methods.go (hunk 3 above)
Comment (rdallman) [inline worker/methods.go]: it's in the docs on dev.iron.io

**Comment by nikhilm** [comment]
LGTM assuming the crypto code was just copy-pasted.

**Comment by rdallman** [comment]
thanks

File: worker/methods.go (hunk 2 above)
Comment (nikhilm) [inline worker/methods.go]: I don't understand? Wouldn't it be something to the effect of `io.Copy(mFileWriter, io.Open(zipName))` where mFileWriter is obtained from CreateFormFile?

Comment (rdallman) [inline worker/methods.go]: eh, doesn't work for zips that easy if memory serves, I think it has something to do with the headers and the naming, have to rewrite all the files out to the multipart file which is rooted and named differently. not worth optimizing really, anyway

**Pull request merged**